kitsune/dashboards/models.py

```diff
@@ -42,29 +26,35 @@ class Meta(object):
     @classmethod
     def reload_period_from_analytics(cls, period, verbose=False):
         """Replace the stats for the given period from Google Analytics."""
-        counts = googleanalytics.pageviews_by_document(*period_dates(period), verbose=verbose)
-        if counts:
-            # Close any existing connections because our load balancer times
-            # them out at 5 minutes and the GA calls take forever.
-            close_old_connections()
-
-            # Delete and remake the rows:
-            # Horribly inefficient until
-            # http://code.djangoproject.com/ticket/9519 is fixed.
-            # cls.objects.filter(period=period).delete()
-
-            # Instead, we use raw SQL!
-            with connection.cursor() as cursor:
-                cursor.execute(
-                    "DELETE FROM dashboards_wikidocumentvisits WHERE period = %s",
-                    [period],
+        with transaction.atomic():
+            # First, let's clear out the previous results for this period.
+            if verbose:
+                log.info(
+                    f"Deleting all stale instances of {cls.__name__} with period = {period}..."
                 )
-            # Now we create them again with fresh data.
-            for doc_id, visits in counts.items():
-                cls.objects.create(document=Document(pk=doc_id), visits=visits, period=period)
-        else:
-            # Don't erase interesting data if there's nothing to replace it:
-            log.warning("Google Analytics returned no interesting data," " so I kept what I had.")
+
+            cls.objects.filter(period=period).delete()
+
+            # Then we can create the fresh results for this period.
+            if verbose:
+                log.info(f"Creating fresh instances of {cls.__name__} with period = {period}...")
+
+            for (locale, slug), visits in googleanalytics.pageviews_by_document(
+                period, verbose=verbose
+            ):
+                try:
+                    with transaction.atomic():
+                        cls.objects.create(
+                            document_id=Subquery(
+                                Document.objects.filter(locale=locale, slug=slug).values("id")
+                            ),
```

Comment:
Would this be a good place for `bulk_create`?

Reply:
Great question, and I should have added a comment about `bulk_create`. I had hoped to use `bulk_create`, but it doesn't handle well the case when the Document instance doesn't exist. It raises an `IntegrityError`, and abandons the entire operation. However, let me see if I can come up with a compromise.

Reply:
I'm really happy with what I came up with. See what you think. I'll test locally but with stage creds on Monday.